lib/puppet/provider/package/dnfmodule.rb

```diff
@@ -0,0 +1,88 @@
+# dnfmodule - A puppet package provider for DNF modules
+#
+# Installing a module:
+#  package { 'postgresql':
+#   provider => 'dnfmodule',
+#   ensure   => '9.6',  # install a specific stream
+#   flavor   => 'client',  # install a specific profile
+# }
+
+
+require 'puppet/provider/package'
+
+Puppet::Type.type(:package).provide :dnfmodule, :parent => :dnf do
+
+  has_feature :installable, :uninstallable, :versionable
+  #has_feature :upgradeable
+  # it's not (yet) feasible to make this upgradeable since module streams don't
+  # always have matching version types (i.e. idm has streams DL1 and client,
+  # other modules have semver streams, others have string streams... we cannot
+  # programatically determine a latest version for ensure => 'latest'
+
+  commands dnf: '/usr/bin/dnf'
+
+  def self.current_version
+    @current_version ||= dnf('--version').split.first
+  end
+
+  def self.prefetch(packages)
+    if Puppet::Util::Package.versioncmp(current_version, '3.0.1') < 0
+      raise Puppet::Error, _("Modules are not supported on DNF versions lower than 3.0.1")
+    end
+    super
+  end
+
+  def self.instances
+    packages = []
+    cmd = "#{command(:dnf)} module list --installed -d 0 -e #{error_level}"
+    execpipe(cmd) do |process|
+      process.each_line do |line|
+        next unless line =~ /\[i\][, ]/  # get rid of non-package lines (including last Hint line)
+        line.gsub!(/\[[de]\]/, '')  # we don't care about default/enabled flags
+        packages << new(
+          name: line.split[0],
+          ensure: line.split[1],
+          flavor: line.split('[i]').first.split.last,  # this is nasty
+          provider: name
+        )
+      end
+    end
+    packages
+  end
+
+  def query
+    self.class.instances.each do |package|
+      return package.properties if @resource[:name] == package.name
+    end
```

Comment:
I generally try to avoid non-local returns, and instead do:

```ruby
pkg = self.class.instances.find do |package|
        @resource[:name] == package.name
      end
pkg ? pkg.properties : nil
```